### ISSUE TITLE
unused_assignments: false negative for dead stores in struct fields

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust#49256]
code like this causes a proper warning
```` rust
fn main() {
	let mut a;
	let b = 4;
	let c = 5;

    	a = b;
    	a = c;

	println!("{}", a);
}
````
=>
````
warning: value assigned to `a` is never read
 --> src/main.rs:6:5
  |
6 |     a = b;
  |     ^
  |
  = note: #[warn(unused_assignments)] on by default
````

however if a is a struct field, I get no warning at all.
```` rust
struct Foo {
    x: i32,
}

fn main() {
	let mut strct = Foo {
	    x: 0,
	};
	let b = 4;
	let c = 5;

  	strct.x = b; // maybe warn here too, since we override the Foo { x: 0 } initialization?
   	strct.x = c; // please warn!

	println!("{}", strct.x);
}
````

playground:  https://play.rust-lang.org/?gist=b4b8f875a0e53f80f24a2247816272ff&version=nightly

I ran into bugs in my code that could have been prevented by warning about this :(

**Comment by Boscop** [comment]
I also ran into a bug where it was a situation like this, also no warning that `a` is never read:
```rust
struct Foo {
    a: i32
}

impl Foo {
    fn bar(&mut self) {
        self.a = 42;
    }
}

fn main() {
    let mut foo = Foo { a: 1 };
    foo.bar();
}
```
https://play.rust-lang.org/?gist=dbe81aa3431e467557fc2bf930827c52&version=stable
